README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
cannonball
==========

cannonball bundles a couple of functions that I use when teaching introductory courses in quantitative methodology and statistics.

Installation
------------

You can install cannonball from Github with:

``` r
library(devtools)
install_github("janhove/cannonball")
```

Functions
---------

``` r
# Load the package
library(cannonball)
```

### `plot_r()`: Draw different scatterplots with the same correlation coefficient

A single correlation coefficient can correspond to any number of scatterplots. `plot_r()` produces 16 such scatterplots for any given Pearson correlation coefficient to drive this point home.

``` r
plot_r(r = 0.5, n = 35)
```

<img src="man/figures/README-example-1.png" width="100%" />

For more info, see `?plot_r`.

Accompanying blog post: [*What data patterns can lie behind a correlation coefficient?*](https://janhove.github.io/teaching/2016/11/21/what-correlations-look-like)

### Walkthroughs

To help students see the connection between an experiment's design and its analysis, I've written two functions.

`walkthrough_p()` guides the user through a completely randomised experiment: Data points are generated and randomly assigned to either the control or intervention condition. Then, the intervention effect is added to the data points in the intervention condition. Finally, the data are analysed using a Student's t-test or, if `pedant = TRUE`, a permutation test.

``` r
# see ?walkthrough_p
walkthrough_p(n = 18, diff = 0.3, sd = 1, pedant = TRUE)
```

`walkthrough_blocking()` works similarly to `walkthrough_p()` but describes a randomised block design: Prior information about the data points is available in the form of a covariate (e.g., a pretest score). This information is used to group participants into 'blocks', and the randomisation is restricted in that one participant per block is assigned to the control and one to the intervention condition. Crucially, the analysis needs to take this restricted randomisation into account.

``` r
# ?walkthrough_blocking
walkthrough_blocking(n = 12, diff = 0.4, sd = 1)
```

An alternative, more powerful, but less easily explained analysis would use the blocking *covariate* as a control variable in a general linear model rather than the blocking *factor*.

### Simulate data for cluster-randomised experiment

The data from experiments in which entire clusters of participants (e.g., classes) are assigned to the experimental conditions can't be analysed in the same way as data from experiments in which the participants are assigned to the conditions individually. `clustered_data()` generates data for a cluster-randomised experiment and can be used to demonstrate the increased Type-I error rate if such data are analysed using t-tests on the individual outcomes.

``` r
# Simulate clustered data with an ICC of 0.3
d <- clustered_data(ICC = 0.3)

# Plot
library(ggplot2)
ggplot(data = d,
       aes(x = reorder(class, outcome, FUN = median),
           y = outcome)) +
  geom_boxplot(outlier.shape = NA) +
  geom_point(shape = 1,
             position = position_jitter(width = 0.1, height = 0)) +
  facet_wrap(~ condition, scales = "free_x") +
  xlab("class")
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />

A covariate can also be created. For more info, see `?clustered_data`.

Accompanying article: Vanhove, Jan. 2015. [Analyzing randomized controlled interventions: Three notes for applied linguists](http://doi.org/10.14746/ssllt.2015.5.1.7). *Studies in Second Language Learning and Teaching* 5(1). 135–152. (See the [correction note](http://pressto.amu.edu.pl/index.php/ssllt/article/view/5827/5895) for this paper.)

Accompanying blog post: [*Experiments with intact groups: spurious significance with improperly weighted t-tests*](https://janhove.github.io/design/2016/02/16/cluster-randomisation-correction).

### Check model assumptions graphically

These functions may be helpful for helping you to judge whether your data conform to the assumptions of your statistical model. By embedding the model's diagnostic plot in a line-up of diagnostic plots of simulated data for which the model's assumptions are literally met, analysts can more easily determine whether any blips in these plots are indicative of assumption violations or whether they can plausibly be accounted for by sampling error/noise. See [Buja et al. (2009)](http://doi.org/10.1098/rsta.2009.0120).

``` r
# Fit model
m <- lm(mpg ~ wt, data = mtcars)

# Create parade/line-up
my_parade <- parade(m)
#> Loading required package: tidyverse
#> ── Attaching packages ────────────────────────────────── tidyverse 1.2.1 ──
#> ✔ tibble  2.1.3     ✔ purrr   0.3.2
#> ✔ tidyr   1.0.0     ✔ dplyr   0.8.3
#> ✔ readr   1.3.1     ✔ stringr 1.4.0
#> ✔ tibble  2.1.3     ✔ forcats 0.4.0
#> ── Conflicts ───────────────────────────────────── tidyverse_conflicts() ──
#> ✖ dplyr::filter() masks stats::filter()
#> ✖ dplyr::lag()    masks stats::lag()

# Check for residual nonlinearities:
# Which plot looks most different from the rest?
lin_plot(my_parade)
#> `geom_smooth()` using method = 'loess' and formula 'y ~ x'
```

<img src="man/figures/README-unnamed-chunk-5-1.png" width="100%" />

``` r

# Reveal position of actual diagnostic plot
reveal(my_parade)
#> The true data are in position 14.
```

Related functions are:

-   `var_plot()`: Check for non-constant variance in the residuals.
-   `norm_qq()`: Check for non-normality of the residuals (quantile--quantile plot).
-   `norm_hist()`: Check for non-normality of the residuals (histogram).
-   `parade_summary()`: Summarise residuals per cell (combination of unique predictor values).

Accompanying article: Vanhove, Jan. 2018. [Checking the assumptions of your statistical model without getting paranoid](https://psyarxiv.com/zvawb/). Preprint on PsyArxiv.


changes
Changes to the cannonball package


2019 - October:

* The number of diagnostic plots in a parade defaults to 20 but can now be changed by the user ('size' parameter in the parade() function).

* In addition to plotting the residuals or absolute residuals against their respective fitted values, they can now also be plotted against their predictor values ('predictor' parameter in the lin_plot() and var_plot() functions).

* The parade() function now offers some support for transformed predictors (e.g., lm(y ~ I(x^2)). The use of poly() and ns() is still not supported, though, as broom::augment() doesn't support this either. See the examples on ?parade.

* parade_summary() now randomly maps the cells in the design to cell numbers.

* clustered_data() now supports measurement error and the generation of a covariate.

* New functions: walkthrough_p() and walkthrough_blocking(). See their help pages for more information.



R/clustered_data.R
#' @title Simulate data for a cluster-randomised experiment
#'
#' @description In educational settings, it's often impossible to randomly assign each individual
#'              students to one of an experiment's conditions. Instead, all students in the same class
#'              or school are assigned to the same condition. This type of random assignment is known as
#'              cluster randomisation. Importantly, if you analyse the data from a cluster-randomised
#'              experiment as though the participants were assigned randomly on an individual basis
#'              (e.g., using a t-test on the individual outcomes), your Type-I error rate (false positive)
#'              goes through the roof. To illustrate this, this function simulates data from a cluster-randomised
#'              experiment. Covariate scores related to the outcome (referred to as pretest scores here)
#'              can also be generated.
#'
#' @details This function works as follows. The within-cluster variance for the participants'
#'          latent pre-intervention skill is set to 1. The within-cluster variance is then computed
#'          on the basis of the desired ICC. Then, pre- and posttest measurements are generated
#'          by adding random normal measurement error to the participants' latent pre-intervention skill.
#'          The variance of the measurement error is computed on the basis of the desired correlation
#'          between the pretest and pre-intervention outcome scores; in this case, the reliabilities
#'          of the pretest and outcome scores are set to be equal.
#'          If no correlation is specified, the measurement error is computed on the basis of the desired
#'          pretest and posttest reliabilities.
#'          Finally, a constant intervention effect is added to the outcome scores of the participants
#'          in the intervention condition.
#'
#'          Note that the square root of the product of the pre- and posttest reliabilities yields
#'          the correlation between them. So if you want to have a correlation between pretest and posttest
#'          scores (without an intervention effect) of 0.7 and you want your pretest to be correlated
#'          at 0.9 to the pre-intervention latent skill levels (i.e., have a reliability of 0.9),
#'          then solve sqrt(0.9*x) = 0.7 for x (x = 0.7^2/0.9 = 0.544).
#'
#' @param n_per_class       The number of participants per class. The first half of the classes are
#'                          assigned to the control condition, the second half to the intervention condition.
#' @param ICC               The intra-class correlation. An ICC of 1 means that the observations in each class are
#'                          identical to each other; an ICC of 0 means that observations within the same class aren't
#'                          any more alike than observations from different classes. Defaults to 0.15.
#' @param effect            The population-level effect of the intervention. Defaults to 0.
#' @param rho_prepost       The desired population-level correlation between the measured pretest scores and outcomes.
#' @param reliability_post  The reliability of the outcomes. Default to 1, meaning there is no measured error on the outcome.
#' @param reliability_pre   The reliability of the pretest scores.
#' @keywords cluster randomisation, simulation
#' @export
#' @references Vanhove, Jan. 2015. Analyzing randomized controlled interventions: Three notes for applied linguists. \emph{Studies in Second Language Learning and Teaching} 5(1). 135-152. \url{http://doi.org/10.14746/ssllt.2015.5.1.7}. (Correction note: \url{http://pressto.amu.edu.pl/index.php/ssllt/article/view/5827/5895}.)
#' @examples
#' # Generate and plot data with ICC = 0
#' d <- clustered_data(ICC = 0)
#' library(tidyverse)
#' ggplot(data = d,
#'        aes(x = reorder(class, outcome, FUN = median),
#'            y = outcome)) +
#'   geom_boxplot(outlier.shape = NA) +
#'   geom_point(shape = 1,
#'              position = position_jitter(width = 0.2)) +
#'   facet_wrap(~ condition, scales = "free_x") +
#'   xlab("outcome")
#'
#' # Generate and plot data with ICC = 0.2
#' d <- clustered_data(ICC = 0.2)
#' ggplot(data = d,
#'        aes(x = reorder(class, outcome, FUN = median),
#'            y = outcome)) +
#'   geom_boxplot(outlier.shape = NA) +
#'   geom_point(shape = 1,
#'              position = position_jitter(width = 0.2)) +
#'   facet_wrap(~ condition, scales = "free_x") +
#'   xlab("outcome")
#'
#' # Generate and plot data with ICC = 0.6
#' d <- clustered_data(ICC = 0.6)
#' ggplot(data = d,
#'        aes(x = reorder(class, outcome, FUN = median),
#'            y = outcome)) +
#'   geom_boxplot(outlier.shape = NA) +
#'   geom_point(shape = 1,
#'              position = position_jitter(width = 0.2)) +
#'   facet_wrap(~ condition, scales = "free_x") +
#'   xlab("outcome")
#'
#' # Simulate 1000 datasets and analyse them using t-test
#' ps <- replicate(1000,
#'                {
#'                  d <- clustered_data(ICC = 0.2)
#'                  return(t.test(outcome ~ condition, data = d)$p.value)
#'                })
#' hist(ps, breaks = seq(0, 1, 0.05))
#' mean(ps < 0.05)
#' # Note the hugely inflated Type-I error rate!
#'
#' # Generate pretest scores with correlation 0.7 to outcome
#' # and with ICC of latent pre-intervention skill of 0.4
#' d <- clustered_data(ICC = 0.4, rho_prepost = 0.7)
#' with(d, cor(pretest, outcome))
#'
#' # ICC = 0.2, pretest reliability = 0.9, posttest reliability = 0.544
#' # and hence correlation = sqrt(0.9*0.544) =~ 0.7:
#' d <- clustered_data(ICC = 0.2, reliability_pre = 0.9, reliability_post = 0.544)
#' with(d, cor(pretest, outcome))

clustered_data <- function(n_per_class = rep(20, 10),
                           ICC = 0.15,
                           effect = 0,
                           rho_prepost = NULL,
                           reliability_post = 1,
                           reliability_pre = NULL) {
  # n_per_class: The cluster sizes. The first half of the classes is assigned to
  #              the intervention condition, the second half the control condition.
  # ICC: The intra-class correlation coefficient.
  # rho_prepost: Covariate not generated if NULL.

  if (min(n_per_class) <= 0 | !isTRUE(all(n_per_class == floor(n_per_class)))) {
    stop("All values in 'n_per_class' should be positive integers.")
  }

  if (ICC < 0 | ICC >= 1) {
    stop("'ICC' should be at least 0 and should not be 1 or higher.")
  }

  if (!is.null(rho_prepost)) {
    if (rho_prepost <= 0 | rho_prepost > 1) {
      stop("'rho_prepost' should either be set to NULL or should be a number between 0 and 1 (0 and 1 are both allowed).")
    }
  }

  if (!is.null(reliability_pre)) {
    if (reliability_pre <= 0 | reliability_pre > 1) {
      stop("'reliability_pre' should be either set to NULL or should be a number between 0 and 1 (1 is allowed, 0 isn't).")
    }
  }

  if (!is.null(reliability_post)) {
    if (reliability_post <= 0 | reliability_post > 1) {
      stop("'reliability_post' should be either set to NULL (in which case 'rho_prepost' should be specified) or should be a number between 0 and 1 (1 is allowed, 0 isn't).")
    }
  }

  # Number of classes
  nr_classes <- length(n_per_class)

  # Compute between-cluster variance. Within-cluster variance is 1.
  var_between <- ICC / (1 - ICC)

  # Generate cluster effects for latent skill without intervention
  Classes <- data.frame(
    class = factor(1:nr_classes),
    condition = factor(c(rep("control", floor(nr_classes/2)),
                         rep("intervention", ceiling(nr_classes/2)))),
    class_effect = rnorm(n = nr_classes, sd = sqrt(var_between))
  )

  # Generate participant effects for latent skill without intervention
  Participants <- data.frame(
    class = rep(Classes$class, times = n_per_class),
    participant_effect = rnorm(n = sum(n_per_class), sd = 1)
  )

  # Combine
  Data <- merge(Classes, Participants, by = "class")

  # Compute outcome and pretest score. The true variance in the latent skill
  # in the entire dataset is the sum of the within- and between-cluster variances.
  true_variance <- var_between + 1

  # If rho_prepost was not set;
  if (is.null(rho_prepost)) {

    # and neither reliability_pre nor reliability_post was set: throw error
    if (is.null(reliability_post) & is.null(reliability_pre)) {
      stop("You need to specify at least 'reliability_post' or 'rho_prepost'.")
    }

    # and reliability_pre was not set either: don't generate pretest scores
    else if (is.null(reliability_pre)) {
      Data$outcome <- Data$class_effect + Data$participant_effect + as.numeric(Data$condition) * effect + rnorm(nrow(Data), sd = sqrt((true_variance*(1-reliability_post))/reliability_post))
      Data$pretest <- NA
    }

    # otherwise generate both pre- and posttest scores
    else {
      # Generate outcomes
      Data$outcome <- Data$class_effect + Data$participant_effect + as.numeric(Data$condition) * effect + rnorm(nrow(Data), sd = sqrt((true_variance*(1-reliability_post))/reliability_post))
      Data$pretest <- Data$class_effect + Data$participant_effect + rnorm(nrow(Data), sd = sqrt((true_variance*(1-reliability_pre))/reliability_pre))
    }
  }

  # if rho_prepost was set
  else if (!is.null(rho_prepost)) {
    # and at least one of the reliabilities was set: generate pre- and posttest scores
    if (!is.null(reliability_post) | !is.null(reliability_pre)) {
      message("'rho_prepost' was set, so the 'reliability_pre' and 'reliability_post' were ignored, and both were set to the 'rho_prepost' value.")
      reliability_post <- reliability_pre <- rho_prepost
      Data$outcome <- Data$class_effect + Data$participant_effect + as.numeric(Data$condition) * effect + rnorm(nrow(Data), sd = sqrt((true_variance*(1-reliability_post))/reliability_post))
      Data$pretest <- Data$class_effect + Data$participant_effect + rnorm(nrow(Data), sd = sqrt((true_variance*(1-reliability_pre))/reliability_pre))
    }

    # if rho_prepost was 0: generate uncorrelated variable
    else if (rho_prepost == 0) {
      # Generate outcomes
      Data$outcome <- Data$class_effect + Data$participant_effect + as.numeric(Data$condition) * effect
      Data$pretest <- rnorm(nrow(Data))
      message("rho_prepost was set to 0. No measurement error was added to the outcome variable, and an uncorrelated pretest score was generated.")
    }
  }

  Data$outcome <- Data$outcome

  # Output
  Data <- subset(Data, select = c("class", "condition", "outcome", "pretest"))
}
